src/mongo/db/storage/csv_file.cpp

```diff
@@ -0,0 +1,338 @@
+
+    auto bsonObj = readBsonObj();
+
+    if (!bsonObj) {
+        return 0;
+    }
+
+    int nRead = bsonObj->objsize();
+    uassert(200000402, "Buff Size too small", nRead <= size);
+    memcpy(data, bsonObj->objdata(), nRead);
+    return nRead;
+}
+
+void CsvFileInput::doClose() {
+    _ifs.close();
+}
+
+bool CsvFileInput::isOpen() const {
+    return _ifs.is_open();
+}
+
+bool CsvFileInput::isGood() const {
+    return _ifs.good();
+}
+
+bool CsvFileInput::isFailed() const {
+    return _ifs.fail();
+}
+
+bool CsvFileInput::isEof() const {
+    return _ifs.eof();
+}
+
+// assuming that header is returned by parseLine, which means, each of its element is
+// name of the field with type name
+Metadata CsvFileInput::getMetadata(const std::vector<std::string>& header) {
+    Metadata ret;
+
+    for (auto& field : header) {
+        int index = field.find('/');
+        int len = field.length() - index - 1;
+
+        auto fieldName = field.substr(0, index);
+        auto typeName = field.substr(index + 1, len);
+        CsvFieldType type;
+
+        if (typeName == "int" || typeName == "int32") {
+            type = CsvFieldType::kInt32;
+        } else if (typeName == "int64" || typeName == "long") {
+            type = CsvFieldType::kInt64;
+        } else if (typeName == "double") {
+            type = CsvFieldType::kDecimal;
+        } else if (typeName == "bool") {
+            type = CsvFieldType::kBool;
+        } else if (typeName == "oid") {
+            type = CsvFieldType::kOid;
+        } else if (typeName == "date") {
+            type = CsvFieldType::kDate;
+        } else if (typeName == "string") {
+            type = CsvFieldType::kString;
+        } else {
+            uasserted(200000401, typeName + " type specified in metadata is not supported");
+        }
+
+        ret.push_back({std::move(fieldName), type});
+    }
+
+    return ret;
+}
+
+template <CsvFieldType T>
+void appendTo(BSONObjBuilder& builder, const std::string& fieldName, const std::string& data);
+
+template <>
+void appendTo<CsvFieldType::kInt32>(BSONObjBuilder& builder,
+                                    const std::string& fieldName,
+                                    const std::string& data) {
+    builder.append(fieldName, stoi(data));
+    if (data.find('.') != std::string::npos) {
+        std::cout << std::endl
+                  << "Lossy conversion of decimal type to whole number type:   Line Number: Field: "
+                  << fieldName << std::endl;
+    }
+}
+
+template <>
+void appendTo<CsvFieldType::kDecimal>(BSONObjBuilder& builder,
+                                      const std::string& fieldName,
+                                      const std::string& data) {
+    builder.append(fieldName, stod(data));
+    int dotIndex = data.find('.');
+    // Warning user of possibly unwanted lossy conversion if provided decimal is too
+    // long
+    if (data.substr(dotIndex + 1, data.length() - dotIndex - 1).length() > 15) {
```

Comment:
This is not right. The number of digits is 15, not the number of fractional digits.